# *_Reconstruction of Small Subunit Ribosomal RNA from High-Throughput Sequencing Data: A Comparative Study of Metagenomics and Total RNA Sequencing_*
This project represents the second chapter of my PhD, in which I compare metagenomics and total RNA-Seq in terms of their Small Subunit (SSU) ribosomal RNA (rRNA) reconstruction completeness and further test the impact of data-processing tools on completeness. The evaluation is based on a microbial mock communtiy. The repo contains scripts to run data-processing pipelines, to determine completeness, and for statistical analysis.

The project is part of the following publication:<br>
<b>Hempel C. A.</b>, Carson S. E. E., Elliott T. A., Adamowicz S. J., and Steinke D. (in review at the journal Methods in Ecology and Evolution): Metagenomics versus total RNA sequencing: most accurate data-processing tools, microbial identification accuracy and perspectives for ecological assessments. Preprint available at [https://doi.org/10.1101/2022.08.26.505493](https://doi.org/10.1101/2022.08.26.505493)

The entire workflow is shown in the following:

<img src="https://github.com/hempelc/metagenomics-vs-totalRNASeq-reference-comparison/blob/main/workflow.png" alt="workflow" width="800"/>
